crates/c-api/src/async.rs

```diff
@@ -0,0 +1,296 @@
+    *mut wasmtime_val_t,
+    usize,
+    &mut Option<Box<wasm_trap_t>>,
+) -> Box<wasmtime_async_continuation_t>;
+
+#[repr(C)]
+pub struct wasmtime_async_continuation_t {
+    pub callback: wasmtime_func_async_continuation_callback_t,
+    pub env: *mut c_void,
+    pub finalizer: Option<extern "C" fn(*mut c_void)>,
+}
+
+unsafe impl Send for wasmtime_async_continuation_t {}
+unsafe impl Sync for wasmtime_async_continuation_t {}
+impl Drop for wasmtime_async_continuation_t {
+    fn drop(&mut self) {
+        if let Some(f) = self.finalizer {
+            f(self.env);
+        }
+    }
+}
+impl Future for wasmtime_async_continuation_t {
+    type Output = ();
+    fn poll(self: Pin<&mut Self>, _cx: &mut Context) -> Poll<Self::Output> {
+        let this = self.get_mut();
+        let cb = this.callback;
+        if cb(this.env) {
+            Poll::Ready(())
+        } else {
+            Poll::Pending
+        }
+    }
+}
+
+pub type wasmtime_func_async_continuation_callback_t = extern "C" fn(*mut c_void) -> bool;
+
+struct CallbackData {
+    env: *mut c_void,
+}
+unsafe impl Send for CallbackData {}
+
+async fn invoke_c_async_callback<'a>(
+    cb: wasmtime_func_async_callback_t,
+    data: CallbackData,
+    mut caller: Caller<'a, crate::StoreData>,
+    params: &'a [Val],
+    results: &'a mut [Val],
+) -> Result<()> {
+    // Convert `params/results` to `wasmtime_val_t`. Use the previous
+    // storage in `hostcall_val_storage` to help avoid allocations all the
+    // time.
+    let mut hostcall_val_storage = mem::take(&mut caller.data_mut().hostcall_val_storage);
+    debug_assert!(hostcall_val_storage.is_empty());
+    hostcall_val_storage.reserve(params.len() + results.len());
+    hostcall_val_storage.extend(params.iter().cloned().map(|p| wasmtime_val_t::from_val(p)));
+    hostcall_val_storage.extend((0..results.len()).map(|_| wasmtime_val_t {
+        kind: WASMTIME_I32,
+        of: wasmtime_val_union { i32: 0 },
+    }));
+    let (params, out_results) = hostcall_val_storage.split_at_mut(params.len());
+
+    // Invoke the C function pointer.
+    // The result will be a continutation which we will wrap in a Future.
+    let mut caller = wasmtime_caller_t { caller };
+    let mut trap = None;
+    let continuation = cb(
+        data.env,
+        &mut caller,
+        params.as_ptr(),
+        params.len(),
+        out_results.as_mut_ptr(),
+        out_results.len(),
+        &mut trap,
+    );
+    continuation.await;
+
+    if let Some(trap) = trap {
+        return Err(trap.error);
+    }
+
+    // Translate the `wasmtime_val_t` results into the `results` space
+    for (i, result) in out_results.iter().enumerate() {
+        unsafe {
+            results[i] = result.to_val();
+        }
+    }
+    // Move our `vals` storage back into the store now that we no longer
+    // need it. This'll get picked up by the next hostcall and reuse our
+    // same storage.
+    let mut v = mem::take(&mut hostcall_val_storage);
```

Comment:
As a small rust-ism the `mem::take` isn't required here and you can say `hostcall_val_storage.truncate(0)` and put it back into the store

Reply:
Thanks! This was vestigial from the refactor :)